libgalois/include/katana/EntityTypeManager.h

```diff
@@ -104,28 +109,47 @@ class EntityTypeManager {
     return new_entity_type_id;
   }
 
-  /// adds a new entity type for the set of atomic types with names \p names
+  /// Adds a new entity type for the set of atomic types with names \p names.
+  /// The returned type is the intersection of the types passed in.
+  ///
   /// \returns the EntityTypeID for the new type
-  EntityTypeID AddNonAtomicEntityType(const std::vector<std::string>& names) {
-    for (auto& name : names) {
-      if (!HasAtomicType(name)) {
-        AddAtomicEntityType(name);
-      }
+  template <typename Container>
+  Result<EntityTypeID> GetOrAddNonAtomicEntityTypeFromStrings(
+      const Container& names) {
+    return GetOrAddNonAtomicEntityType(KATANA_CHECKED(GetOrAddEntityTypeIDs(names)));
+  }
+
+  /// Adds a new entity type for the set of atomic types.
+  /// The returned type is the intersection of the types passed in.
+  ///
+  /// \returns the EntityTypeID for the new type
+  Result<EntityTypeID> GetOrAddNonAtomicEntityType(
+      const SetOfEntityTypeIDs& type_id_set) {
+    auto existing_id = std::find(
```

Comment:
Adding types is not a high-performance operation, but this is a linear search?  Is that ok?

Reply:
That's a reasonable question. I made a space/time trade-off here in favor of space to avoid adding another table. I also judged that a linear search is only `O(n types)` and we do not expect many types (even including intersection types as we are here).

Previously there was no check like this, but I'm 99% sure that we were breaking uniqueness of types before and that you could end up with multiple type IDs having the same set of supertypes. I think this should be avoided and I think it would cause problems even with the current code.

Reply:
Yeah, I worried about uniqueness as well.
I think this is a fine decision, but I would add a comment saying this is a linear search that assumes a small total number of types.